src/openms/source/FILTERING/SMOOTHING/FastLowessSmoothing.cpp

```diff
@@ -57,269 +57,350 @@
+  *  # radius will not decrease by move right
+  *                         nleft = nleft+1
+  *                         nright = nright+1
+  *                         }
+  *                 call lowest(x,y,n,x(i),ys(i),nleft,nright,res,iter>1,rw,ok)
+  *  # fitted value at x(i)
+  *                 if (!ok) ys(i) = y(i)
+  *  # all weights zero - copy over value (all rw==0)
+  *                 if (last<i-1) { # skipped points -- interpolate
+  *                         denom = x(i)-x(last)    # non-zero - proof?
+  *                         for(j=last+1; j<i; j=j+1){
+  *                                 alpha = (x(j)-x(last))/denom
+  *                                 ys(j) = alpha*ys(i)+(1.0-alpha)*ys(last)
+  *                                 }
+  *                         }
+  *                 last = i        # last point actually estimated
+  *                 cut = x(last)+delta     # x coord of close points
+  *                 for(i=last+1; i<=n; i=i+1){     # find close points
+  *                         if (x(i)>cut) break     # i one beyond last pt within cut
+  *                         if(x(i)==x(last)){      # exact match in x
+  *                                 ys(i) = ys(last)
+  *                                 last = i
+  *                                 }
+  *                         }
+  *                 i=max0(last+1,i-1)
+  *  # back 1 point so interpolation within delta, but always go forward
+  *                 } until(last>=n)
+  *         do i = 1,n      # residuals
+  *                 res(i) = y(i)-ys(i)
+  *         if (iter>nsteps) break  # compute robustness weights except last time
+  *         do i = 1,n
+  *                 rw(i) = abs(res(i))
+  *         call sort(rw,n)
+  *         m1 = 1+n/2; m2 = n-m1+1
+  *         cmad = 3.0*(rw(m1)+rw(m2))      # 6 median abs resid
+  *         c9 = .999*cmad; c1 = .001*cmad
+  *         do i = 1,n {
+  *                 r = abs(res(i))
+  *                 if(r<=c1) rw(i)=1.      # near 0, avoid underflow
+  *                 else if(r>c9) rw(i)=0.  # near 1, avoid underflow
+  *                 else rw(i) = (1.0-(r/cmad)**2)**2
+  *                 }
+  *         }
+  *  return
+  *  end
+
+  */
+
+  /// Templated lowess class, call with template container (can be anything
+  /// that supports random access)
+  template <typename ContainerType, typename ValueType>
+  class TemplatedLowess
+  {
 
-  a = 0.0; // sum of weights
+    inline ValueType pow2(ValueType x) { return x * x;  }
+    inline ValueType pow3(ValueType x) { return x * x * x;  }
 
-  // compute weights (pick up all ties on right)
-  for (j = nleft; j < (long)n; j++) 
-  {
+    ///Return the median of a sequence of numbers defined by the random
+    ///access iterators begin and end.  The sequence must not be empty
+    ///(median is undefined for an empty set).
+    ///
+    ///The numbers must be convertible to double.
+    template <class RandAccessIter>
```

Comment:
mhh... there is an implementation of median in OpenMS already (http://ftp.mi.fu-berlin.de/pub/OpenMS/documentation/html/group__MathFunctionsStatistics.html#ga237d44c9a46ea3092485ff369795a053)

I can see that you need an independent one if the project should be standalone.. but do you see any way to avoid this?!